timelock-corruption-detection/src/main/java/com/palantir/timelock/history/PaxosLogHistoryProvider.java

```diff
@@ -61,16 +77,153 @@ private Long getOrInsertVerificationState(NamespaceAndUseCase namespaceAndUseCas
+            NamespaceAndUseCase namespaceAndUseCase,
+            List<ConsolidatedPaxosHistoryOnSingleNode> historyLogsFromRemotes) {
+        return historyLogsFromRemotes
+                .stream()
+                .map(history -> history.getRecordForNamespaceAndUseCase(namespaceAndUseCase))
+                .collect(Collectors.toList());
+    }
+
+    private List<ConsolidatedLearnerAndAcceptorRecord> combineLocalAndRemoteHistoryLogs(
+            ConsolidatedLearnerAndAcceptorRecord consolidatedLocalRecord,
+            List<ConsolidatedLearnerAndAcceptorRecord> consolidatedRemoteRecords) {
+        return ImmutableList
+                .<ConsolidatedLearnerAndAcceptorRecord>builder()
+                .addAll(consolidatedRemoteRecords)
+                .add(consolidatedLocalRecord)
+                .build();
+    }
+
+    private ConsolidatedPaxosHistoryOnSingleNode buildRecordFromRemoteResponse(
+            PaxosHistoryOnRemote historyOnRemote) {
+
+        Map<NamespaceAndUseCase, List<PaxosLogWithAcceptedAndLearnedValues>> namespaceWisePaxosLogs
+                = historyOnRemote.getLogs()
+                .stream()
+                .collect(Collectors.toMap(
+                        LogsForNamespaceAndUseCase::getNamespaceAndUseCase,
+                        LogsForNamespaceAndUseCase::getLogs)
+                );
+
+        return ConsolidatedPaxosHistoryOnSingleNode.of(KeyedStream
+                .stream(namespaceWisePaxosLogs)
+                .map(this::getConsolidatedLearnerAndAcceptorRecordFromRemotePaxosLogs)
+                .collectToMap());
+    }
+
+    private ConsolidatedLearnerAndAcceptorRecord getConsolidatedLearnerAndAcceptorRecordFromRemotePaxosLogs(
+            NamespaceAndUseCase unused,
+            List<PaxosLogWithAcceptedAndLearnedValues> remoteLogs) {
+        return ConsolidatedLearnerAndAcceptorRecord
+                .of(getSequenceWiseLearnedAndAcceptedValuesFromRemoteLogs(remoteLogs));
+    }
+
+    private Map<Long, LearnedAndAcceptedValue> getSequenceWiseLearnedAndAcceptedValuesFromRemoteLogs(
+            List<PaxosLogWithAcceptedAndLearnedValues> remoteLogs) {
+        return remoteLogs
+                .stream()
+                .collect(Collectors.toMap(
+                        PaxosLogWithAcceptedAndLearnedValues::getSeq,
+                        remoteLog -> ImmutableLearnedAndAcceptedValue.of(
+                                remoteLog.getPaxosValue(),
+                                remoteLog.getAcceptedState())));
+    }
+
+    private PaxosHistoryOnRemote fetchHistoryFromRemote(List<HistoryQuery> historyQueries,
+            TimeLockPaxosHistoryProvider remote) {
+        try {
+            return remote.getPaxosHistory(AUTH_HEADER, historyQueries);
+        } catch (Exception exception) {
+            log.warn("The remote failed to provide the history,"
+                    + "we cannot perform corruption checks without history from all nodes.", exception);
```

Comment:
nit: space after comma
```suggestion
                    + " we cannot perform corruption checks without history from all nodes.", exception);
```